### PR TITLE
http2: specify default TLS options for http2 client connection.

### LINKED ISSUE #59: a better default configuration for TLS connection of http2
It seems the TLS socket is initialized as `tls.connect(port, host, options)` in `http2.connect`. Some existing HTTP2 servers require some TLS options to set up the connection properly. More specifically:
- servername (for Server Name Indication)
- ALPN / NPN, specifying 'h2'

I believe those should be set by defaults, then passing a URL string will just work.

Also, maybe it's better to use `secureConnect` event instead of `connect` event for the TLS sockets for setting up a session -- so that it won't go into the session if some TLS setup failure happens.

### PR DESCRIPTION
**Pull request by jmuk**
fixes: https://github.com/nodejs/http2/issues/59

### CODE REVIEW

File: lib/internal/http2/core.js [hunk 1]
```diff
@@ -1235,6 +1235,11 @@ function connect(authority, options, listener) {
       socket = net.connect(port, host);
       break;
     case 'https:':
+      options = util._extend({
+        servername: host,
+        ALPNProtocols: ['h2'],
+        NPNProtocols: ['h2']
+      }, options);
```
Comment (mcollina) [inline lib/internal/http2/core.js]: I would prefer `Object.assign`

**Review by mcollina — CHANGES_REQUESTED** [review]
can you also add a unit test for this?

**Comment by jasnell** [comment]
Had a different approach in mind for fixing this: 

```js
diff --git a/lib/internal/http2/core.js b/lib/internal/http2/core.js
index 6afc87e6f7..052513289f 100755
--- a/lib/internal/http2/core.js
+++ b/lib/internal/http2/core.js
@@ -1077,6 +1077,7 @@ function initializeOptions(options) {
   options = options || {};
   if (typeof options !== 'object')
     throw new TypeError('options must be an object');
+  options = Object.create(options);
   options.allowHalfOpen = true;
   options.settings = options.settings || {};
   if (typeof options.settings !== 'object')
@@ -1084,10 +1085,12 @@ function initializeOptions(options) {
   return options;
 }

-function initializeTLSOptions(options) {
+function initializeTLSOptions(options, host) {
   options = initializeOptions(options);
   options.ALPNProtocols = ['hc', 'h2'];
   options.NPNProtocols = ['hc', 'h2'];
+  if (host !== undefined)
+    options.servername = host;
   return options;
 }

@@ -1235,7 +1238,7 @@ function connect(authority, options, listener) {
       socket = net.connect(port, host);
       break;
     case 'https:':
-      socket = tls.connect(port, host, options);
+      socket = tls.connect(port, host, initializeTLSOptions(options, host));
       break;
     default:
       throw new TypeError(`protocol "${protocol}" in unsupported.`);
james@ubuntu:~/node/http2-jasnell$
```

This allows us to have a central place for setting the TLS options for both server and client.

**Comment by jmuk** [comment]
Thank you for the review!  Added a test case, and fixed the client code for your latter suggestion.

**Comment by jmuk** [comment]
By the way -- where does the ALPN name `'hc'` come from? I can't find the document referring that value -- http2 spec would suggest to use `h2` only for TLS context, I suppose (see https://http2.github.io/http2-spec/#rfc.section.3.1).

**Comment by jasnell** [comment]
I think it's an old pre-standard value. One of the browser impls was using it. Whether or not we need to keep it remains to be seen

File: test/parallel/test-http2-create-client-secure-session.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        const jsonData = JSON.parse(data);
+        assert.strictEqual(jsonData.servername, 'localhost');
+        assert(jsonData.alpnProtocol === 'h2' || jsonData.alpnProtocol === 'hc');
+        if (--expected === 0) {
+          server.close();
+          client.socket.destroy();
+        }
+      }));
+      req.end();
+    }
```
Comment (mcollina) [inline test/parallel/test-http2-create-client-secure-session.js]: why are you repeating this test 10 times?

Comment (jmuk) [reply test/parallel/test-http2-create-client-secure-session.js]: I borrowed the structure from `test/parallel/test-http2-create-client-session.js` which also repeats 10 times (https://github.com/nodejs/http2/blob/master/test/parallel/test-http2-create-client-session.js), I think it's better to have the same structure here. Should I change here, or both?

Comment (jasnell) [reply test/parallel/test-http2-create-client-secure-session.js]: the calling ten times in the other test is an artifact from trying to trace down a bug I was working through. I actually need to refactor that test to pull that back out :-)

**Comment by jmuk** [comment]
Removed the count, thank you for the background info.

**Review by jasnell — APPROVED** [review]

**Comment by jasnell** [comment]
@mcollina ... ping

**Review by mcollina — APPROVED** [review]

**Comment by mcollina** [comment]
LGTM

**Comment by jasnell** [comment]
Landed!

**Pull request closed without merging**